File: java/proto/src/main/java/com/pravala/protocol/Serializable.java
/*
 *  Copyright 2019 Carnegie Technologies
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package com.pravala.protocol;

import com.pravala.protocol.auto.ErrorCode;

import java.io.IOException;
import java.io.OutputStream;

/**
 * A class that contains base methods for serialization and deserialization
 */
public abstract class Serializable
{
    /**
     * Serializes content of the message
     *
     * It just appends content to the buffer.
     * First the serializeData() from inherited object is called.
     * Next, all present, local fields are serialized -
     * either by serializing them directly, or by calling their serializeData()
     * This function does not encode base message's length.
     * To encode a complete message you need to call serialize()
     * from the base message class (and that is the only function that should
     * be used in most cases - usually there is no need to call serializeData() directly)
     * @param writeTo Where to serialize data
     * @throws IOException, CodecException
     */
    public void serializeData ( OutputStream writeTo ) throws IOException, CodecException
    {
        // If we 'define' and alias, we are safe to do this first and call 'serialize'
        // in our ancestor later (which could, potentially, define the storage type itself
        // overwriting our value). However, the protocol compiler should not allow
        // storage (aliased) types to be defined (and vice versa).
        setupDefines();

        validate();

        serializeAllFields ( writeTo );
    }

    /**
     * Deserializes data from the buffer
     *
     * First the deserialize() from inherited object is called.
     * Next, all present, local fields are deserialized -
     * either by deserializing them directly, or by calling their deserialize()
     * @param readBuffer The buffer to deserialize the data from. Its offset is NOT modified.
     * @return True if the message has been successfully deserialized,
     *         False if an unknown field was found, and throws an exception on errors
     * @throws CodecException
     */
    public boolean deserializeData ( ReadBuffer readBuffer ) throws CodecException
    {
        if ( readBuffer == null || readBuffer.getOffset() < 0
             || readBuffer.getSize() < 1 || readBuffer.getOffset() > readBuffer.getSize() )
        {
            throw new CodecException ( ErrorCode.InvalidParameter );
        }

        clear();

        boolean ret = true;

        int offset = readBuffer.getOffset();

        while ( offset < readBuffer.getSize() )
        {
            ReadBuffer tmpBuf = new ReadBuffer ( readBuffer.getBuffer(), offset, readBuffer.getSize() );

            Codec.FieldHeader hdr = Codec.readFieldHeader ( tmpBuf );

            // The new offset (it doesn't affect or depend on the deserializeField!):
            offset = tmpBuf.getOffset() + hdr.fieldSize;

            if ( offset > tmpBuf.getSize() )
            {
                throw new CodecException ( ErrorCode.IncompleteData );
            }

            tmpBuf.setSize ( offset );

            if ( !deserializeField ( hdr, tmpBuf ) )
            {
                ret = false;
            }
        }

        // Regardless of the 'ret' value we want to check if everything is OK
        validate();

        return ret;
    }

    /**
     * Clears the content
     *
     * All fields will either be set to their default values (or 0 if not set) or their clear()
     * method will be called and they will be set as not present.
     */
    public abstract void clear();

    /**
     * Checks validity of the data
     *
     * Checks if all required fields in this and all inherited objects are present and have legal values.
     * If this is used by external code on messages that are to be sent
     * (NOT on received ones!) it is probably a good idea to call setupDefines() first.
     *
     * @throws CodecException
     */
    public abstract void validate() throws CodecException;

    /**
     * Sets the values of all the fields 'defined' by this and all inherited objects (if any)
     */
    public abstract void setupDefines();

    /**
     * Serializes all fields to the buffer. It is called by serializeData.
     * @param writeTo Where to serialize data
     * @throws IOException, CodecException
     */
    protected abstract void serializeAllFields ( OutputStream writeTo ) throws IOException, CodecException;

    /**
     * Deserializes a single field from the buffer.
     * @param hdr Field's header description.
     * @param readBuffer The buffer to deserialize the data from. The offset should be pointing at the first byte
     *                   of the actual payload (just after the header).
     *                   This function may modify this buffer's offset. It may also be invalid after calling it,
     *                   or in the middle of the field's payload.
     * @return True if the field has been successfully deserialized,
     *         False if an unknown field was found, and throws an exception on errors
     * @throws CodecException
     */
    protected abstract boolean deserializeField ( Codec.FieldHeader hdr, ReadBuffer readBuffer ) throws CodecException;
}

File: csharp/Protocol/Serializable.cs
/*
 *  Copyright 2019 Carnegie Technologies
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

//// #define DEBUG2

namespace Pravala.Protocol
{
    using System.Diagnostics;

    /// <summary>
    /// A base class that should be inherited by all messages.
    /// </summary>
    public abstract class Serializable
    {
        /// <summary>
        /// Clears all the fields.
        /// </summary>
        public virtual void ClearMessage()
        {
        }

        /// <summary>
        /// Validates the content of the object.
        /// </summary>
        /// <exception cref="ProtoException">If the message is incorrect.</exception>
        public virtual void ValidateMessage()
        {
        }

        /// <summary>
        /// Configures all "defined" fields.
        /// </summary>
        /// Checks if all required fields in this and all inherited objects are present and have legal values.
        /// If this is used by external code on messages that are to be sent
        /// it is probably a good idea to call setupDefines() first.
        public virtual void SetupMessageDefines()
        {
        }

        /// <summary>
        /// Deserializes the data from the buffer.
        /// </summary>
        /// <param name="buffer">Buffer to deserialize data from.</param>
        /// <param name="offset">Starting offset.</param>
        /// <param name="length">The number of bytes that can be read (starting at offset, not 0).</param>
        /// <returns>
        /// <c>true</c>, if all fields in this message were known,
        /// <c>false</c> if some of the fields were unknown.
        /// </returns>
        /// <exception cref="ProtoException">
        /// On error or if the message doesn't validate after being deserialized.
        /// </exception>
        internal virtual bool DeserializeMessageData ( byte[] buffer, uint offset, uint length )
        {
            this.ClearMessage();

            uint maxSize = offset + length;

            if ( buffer == null || maxSize > buffer.Length )
            {
                throw new ProtoException (
                        ProtoException.ErrorCodes.InvalidParameter,
                        "No buffer provided, or incorrect buffer size" );
            }

            bool ret = true;

            uint desOffset = offset;

            while ( offset < maxSize )
            {
                uint fieldId;
                uint payloadSize;
                Codec.WireTypes wireType;
                uint fieldOffset = offset;

#if DEBUG2
                Debug.WriteLine ( "Field; Offset: " + offset + "; MaxSize: " + maxSize );
#endif

                if ( !Codec.ReadHeader ( buffer, ref offset, maxSize, out fieldId, out payloadSize, out wireType ) )
                {
                    // There is not enough data. In this function we should have enough data!
                    throw new ProtoException (
                            ProtoException.ErrorCodes.InvalidDataSize,
                            "Not enough data to decode next field's header" );
                }

#if DEBUG2
                Debug.WriteLine ( "FieldId: " + fieldId + "; WireType: " + wireType + "; Offset: "
                        + offset + "; PayloadSize: " + payloadSize + "; MaxSize: " + maxSize );
#endif

                if ( offset + payloadSize > maxSize )
                {
                    // There was enough data to decode the header,
                    // but the payload size in this header describes data that we don't have...
                    throw new ProtoException (
                            ProtoException.ErrorCodes.InvalidDataSize,
                            "Not enough data to decode next field's payload; Field Id: " + fieldId
                            + "; WireType: " + ( ( int ) wireType )
                            + "; PayloadSize: " + payloadSize + "; DesOffset: " + desOffset
                            + "; FieldOffset: " + fieldOffset + "; CurOffset: " + offset
                            + "; MaxSize: " + maxSize + "; Length: " + length
                            + "; BufSize: " + buffer.Length );
                }

                if ( !this.DeserializeMessageField ( buffer, offset, fieldId, payloadSize, wireType ) )
                {
                    // Unknown field
                    ret = false;
                }

                // If we are still here (no exception),
                // it means that the field was handled properly (even if it was unknown)!
                offset += payloadSize;
            }

            this.ValidateMessage();

            return ret;
        }

        /// <summary>
        /// Serializes the object into the buffer.
        /// </summary>
        /// It first configures the defined values and then it tries to validate the message.
        /// If it validates correctly it is serialized into the buffer.
        /// <param name="buffer">Buffer to serialize data to.</param>
        /// <exception cref="ProtoException">On error or if the message doesn't validate.</exception>
        internal void SerializeMessageData ( Buffer buffer )
        {
            this.SetupMessageDefines();
            this.ValidateMessage();

            if ( buffer == null )
            {
                throw new ProtoException ( ProtoException.ErrorCodes.InvalidParameter, "No buffer provided" );
            }

            this.SerializeMessageFields ( buffer );
        }

        /// <summary>
        /// Deserializes a single field.
        /// </summary>
        /// <param name="buffer">Buffer to deserialize the payload from.</param>
        /// <param name="offset">Offset in the buffer where the payload starts.</param>
        /// <param name="fieldId">Field identifier.</param>
        /// <param name="payloadSize">Payload size.</param>
        /// <param name="wireType">The wire type used by the field.</param>
        /// <returns><c>true</c>, if field was deserialized, <c>false</c> otherwise (if it was unknown).</returns>
        /// <exception cref="ProtoException">If the decoding failed.</exception>
        protected virtual bool DeserializeMessageField (
                byte[] buffer,
                uint offset,
                uint fieldId,
                uint payloadSize,
                Codec.WireTypes wireType )
        {
            if ( buffer == null || offset + payloadSize > buffer.Length )
            {
                throw new ProtoException (
                        ProtoException.ErrorCodes.InvalidParameter,
                        "Invalid buffer, or incorrect offset/payload size" );
            }

            return false;
        }

        /// <summary>
        /// Serializes all fields into the buffer.
        /// </summary>
        /// <param name="buffer">Buffer to serialize data to.</param>
        protected virtual void SerializeMessageFields ( Buffer buffer )
        {
        }
    }
}
